sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java

```diff
@@ -336,4 +336,62 @@ public UnsafeArrayData copy() {
     arrayCopy.pointTo(arrayDataCopy, Platform.BYTE_ARRAY_OFFSET, sizeInBytes);
     return arrayCopy;
   }
+
+  public int[] toPrimitiveIntArray() {
+    int[] result = new int[numElements];
+    Platform.copyMemory(baseObject, baseOffset + 4 + 4 * numElements,
+      result, Platform.INT_ARRAY_OFFSET, 4 * numElements);
+    return result;
+  }
+
+  public double[] toPrimitiveDoubleArray() {
+    double[] result = new double[numElements];
+    Platform.copyMemory(baseObject, baseOffset + 4 + 4 * numElements,
+      result, Platform.DOUBLE_ARRAY_OFFSET, 8 * numElements);
+    return result;
+  }
+
+  public static UnsafeArrayData fromPrimitiveArray(int[] arr) {
+    int offsetRegionSize = 4 * arr.length;
+    int valueRegionSize = 4 * arr.length;
+    int totalSize = 4 + offsetRegionSize + valueRegionSize;
+    byte[] data = new byte[totalSize];
```

Comment:
After think about it, we can't support very long array in `UnsafeArrayData`, as the offset region only use 4 bytes to encode the value offset. I'm going to add a validation so that we won't overflow if given a big array.

Actually we can improve the unsafe format for array data, which is similar to unsafe row:
```
[null bits] [values] [variable length portion]
```

One difference is that, the `values` region in unsafe row is 8-byte per field, but for array, it should depend on the data size, e.g. 1 byte for boolean. Then we can just memory copy primitive array into `values` region and no need to care about the offset region.